astropy/coordinates/sky_coordinate_parsers.py

```diff
@@ -257,69 +252,64 @@ def _parse_coordinate_data(frame, args, kwargs):
             )
         )
 
+    if len(args) > 3:
+        raise ValueError(
+            f"Must supply no more than three positional arguments, got {len(args)}"
+        )
+
     # Finally deal with the unnamed args.  This figures out what the arg[0]
     # is and returns a dict with appropriate key/values for initializing
     # frame class. Note that differentials are *never* valid args, only
     # kwargs.  So they are not accounted for here (unless they're in a frame
     # or SkyCoord object)
-    if args:
-        if len(args) == 1:
-            # One arg which must be a coordinate.  In this case coord_kwargs
-            # will contain keys like 'ra', 'dec', 'distance' along with any
-            # frame attributes like equinox or obstime which were explicitly
-            # specified in the coordinate object (i.e. non-default).
-            _skycoord_kwargs, _components = _parse_coordinate_arg(
-                args[0], frame, units, kwargs
-            )
-
-            # Copy other 'info' attr only if it has actually been defined.
-            if "info" in getattr(args[0], "__dict__", ()):
-                info = args[0].info
-
-        elif len(args) <= 3:
-            _skycoord_kwargs = {}
-            _components = {}
+    _skycoord_kwargs = {}
+    _components = {}
+    if len(args) > 1:
+        for arg, (frame_attr_name, repr_attr_name), unit in zip(
+            args, frame.representation_component_names.items(), units
+        ):
+            attr_class = frame.representation_type.attr_classes[repr_attr_name]
+            _components[frame_attr_name] = attr_class(arg, unit=unit)
+
+    elif len(args) == 1:
+        # One arg which must be a coordinate.  In this case coord_kwargs
+        # will contain keys like 'ra', 'dec', 'distance' along with any
+        # frame attributes like equinox or obstime which were explicitly
+        # specified in the coordinate object (i.e. non-default).
+        _skycoord_kwargs, _components = _parse_coordinate_arg(
+            args[0], frame, units, kwargs
```

Comment:
Not your change, but why is `kwargs` pass through when it is guaranteed to be empty (otherwise error at l.232)?

Reply:
Turns out `_parse_coordinate_arg()` doesn't need its fourth argument at all, but I'm not sure if I should be editing the signature of `_parse_coordinate_arg()` in a pull request that was only meant to shorten `_parse_coordinate_data()`.

Reply:
Yes, I also wondered if it wasn't better to keep this PR just focused on doing the one thing well...

Indeed, would be happy to merge, except that readthedocs seems stuck.